module/zfs/zfs_ioctl.c

```diff
@@ -2819,8 +2777,49 @@ zfs_ioc_inherit_prop(zfs_cmd_t *zc)
 			return (SET_ERROR(EINVAL));
 	}
 
-	/* property name has been validated by zfs_secpolicy_inherit_prop() */
-	return (dsl_prop_inherit(zc->zc_name, zc->zc_value, source));
+	if (prop == ZPROP_INVAL) {
+		if (!zfs_prop_user(propname))
+			return (SET_ERROR(EINVAL));
+
+		type = PROP_TYPE_STRING;
+	} else if (prop == ZFS_PROP_VOLSIZE || prop == ZFS_PROP_VERSION) {
+		return (SET_ERROR(EINVAL));
+	} else {
+		type = zfs_prop_get_type(prop);
+	}
+
+	/*
+	 * zfs_prop_set_special() expects properties in the form of an
+	 * nvpair with type info.
+	 */
+	dummy = fnvlist_alloc();
+
+	switch (type) {
+	case PROP_TYPE_STRING:
+		VERIFY(0 == nvlist_add_string(dummy, propname, ""));
+		break;
+	case PROP_TYPE_NUMBER:
+	case PROP_TYPE_INDEX:
+		VERIFY(0 == nvlist_add_uint64(dummy, propname, 0));
+		break;
+	default:
+		err = SET_ERROR(EINVAL);
+		goto errout;
+	}
+
+	pair = nvlist_next_nvpair(dummy, NULL);
+	if (pair == NULL) {
+		err = SET_ERROR(EINVAL);
+	} else {
+		err = zfs_prop_set_special(zc->zc_name, source, pair);
+		if (err == -1) /* property is not "special", needs handling */
+			err = dsl_prop_inherit(zc->zc_name, zc->zc_value,
```

Comment:
> zfs_set_prop_special() is called in your code for all kinds of properties, including user properties, special, and non-special inheritable properties, which seems totally superficial and unnecessary. Incidental to that, there are nvlists created and freed, arguments prepared for zfs_set_prop_special() only to have it return -1 and generate a bunch of heat and no other effect.

> To summarize, I don't like the way you have restructured the code paths. I find it much harder to understand what's going on, and the hole bunch of no-op processing occurring for no reason whatsoever is something I that prefer to avoid.

Can you propose a prototype (even pseudocode is fine) that would allow us to avoid unnecessary processing (no-ops, nvlist alloc/free, calling `zfs_set_prop_special()` for non-"special" properties...) but still retain the same functionality? Thanks